File: termux_playwright/waf.py
"""WAF and bot challenge evaluator for termux-playwright.

Detects Cloudflare Turnstile, Cloudflare Managed Challenge, hCaptcha, and reCAPTCHA.
Provides coordinated auto-resolution via HumanMouse interaction models.
"""

import asyncio
import enum
import logging
import random
from typing import Optional, Any
from .physics import HumanMouse

logger = logging.getLogger(__name__)


class WafChallengeType(str, enum.Enum):
    """Known web application firewall bot challenge types."""
    CLOUDFLARE_TURNSTILE = "cloudflare_turnstile"
    CLOUDFLARE_MANAGED = "cloudflare_managed"
    HCAPTCHA = "hcaptcha"
    RECAPTCHA = "recaptcha"
    NONE = "none"


class TurnstileEvaluator:
    """Evaluates page state for Cloudflare Turnstile / WAF challenges and executes human interaction."""

    CLOUDFLARE_SELECTORS = [
        "iframe[src*='challenges.cloudflare.com']",
        "iframe[src*='cf-turnstile']",
        "#cf-stage",
        "#turnstile-wrapper",
        "div[class*='cf-turnstile']",
    ]

    HCAPTCHA_SELECTORS = [
        "iframe[src*='hcaptcha.com']",
        "div.h-captcha",
    ]

    RECAPTCHA_SELECTORS = [
        "iframe[src*='google.com/recaptcha']",
        "iframe[src*='recaptcha']",
        "div.g-recaptcha",
    ]

    @classmethod
    def _get_first(cls, loc: Any) -> Any:
        if hasattr(loc, "first"):
            first_val = getattr(loc, "first")
            return first_val() if callable(first_val) else first_val
        return loc

    @classmethod
    async def detect_challenge(cls, page: Any) -> WafChallengeType:
        """Detects the presence of WAF challenge widgets on the current page."""
        try:
            for sel in cls.CLOUDFLARE_SELECTORS:
                loc = cls._get_first(page.locator(sel))
                if await loc.count() > 0:
                    return WafChallengeType.CLOUDFLARE_TURNSTILE

            title = (await page.title()).lower() if hasattr(page, "title") else ""
            if "just a moment" in title or "attention required" in title:
                return WafChallengeType.CLOUDFLARE_MANAGED

            for sel in cls.HCAPTCHA_SELECTORS:
                loc = cls._get_first(page.locator(sel))
                if await loc.count() > 0:
                    return WafChallengeType.HCAPTCHA

            for sel in cls.RECAPTCHA_SELECTORS:
                loc = cls._get_first(page.locator(sel))
                if await loc.count() > 0:
                    return WafChallengeType.RECAPTCHA

        except Exception as e:
            logger.debug(f"Challenge detection error: {e}")

        return WafChallengeType.NONE

    @classmethod
    async def solve_turnstile(
        cls,
        page: Any,
        human_mouse: Optional[HumanMouse] = None,
        timeout: float = 12.0,
    ) -> bool:
        """Locates the Cloudflare Turnstile iframe and clicks the verification checkbox using human mouse movement."""
        mouse = human_mouse or HumanMouse(page)
        deadline = asyncio.get_event_loop().time() + timeout

        while asyncio.get_event_loop().time() < deadline:
            try:
                for frame_sel in cls.CLOUDFLARE_SELECTORS:
                    frame_loc = cls._get_first(page.locator(frame_sel))
                    if await frame_loc.count() > 0:
                        box = await frame_loc.bounding_box()
                        if box and box["width"] > 20 and box["height"] > 20:
                            # Target the checkbox inside the Turnstile widget (approx 30px from left, 50% height)
                            target_x = box["x"] + min(36.0, box["width"] * 0.2) + random.uniform(-4, 4)
                            target_y = box["y"] + (box["height"] * 0.5) + random.uniform(-4, 4)

                            await mouse.click(
                                target=(target_x, target_y),
                                steps=random.randint(25, 38),
                                jitter=True,
                                overshoot=True,
                            )
                            # Wait for challenge resolution
                            await asyncio.sleep(0.5)
                            return True
            except Exception as e:
                logger.debug(f"Solve Turnstile attempt error: {e}")

            await asyncio.sleep(0.1)

        return False

File: lib/waf.js
/**
 * Termux-Playwright Node.js WAF & Bot Challenge Evaluator
 * Cloudflare Turnstile, hCaptcha, and reCAPTCHA auto-detection and resolution.
 * @license MIT
 */

'use strict';

const { HumanMouse } = require('./physics');

const WafChallengeType = {
    CLOUDFLARE_TURNSTILE: 'cloudflare_turnstile',
    CLOUDFLARE_MANAGED: 'cloudflare_managed',
    HCAPTCHA: 'hcaptcha',
    RECAPTCHA: 'recaptcha',
    NONE: 'none'
};

const CLOUDFLARE_SELECTORS = [
    "iframe[src*='challenges.cloudflare.com']",
    "iframe[src*='cf-turnstile']",
    '#cf-stage',
    '#turnstile-wrapper',
    "div[class*='cf-turnstile']"
];

const HCAPTCHA_SELECTORS = [
    "iframe[src*='hcaptcha.com']",
    'div.h-captcha'
];

const RECAPTCHA_SELECTORS = [
    "iframe[src*='google.com/recaptcha']",
    "iframe[src*='recaptcha']",
    'div.g-recaptcha'
];

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

class TurnstileEvaluator {
    /**
     * Evaluates whether a known bot challenge widget is on the page.
     * @param {any} page
     * @returns {Promise<string>}
     */
    static async detectChallenge(page) {
        try {
            for (const sel of CLOUDFLARE_SELECTORS) {
                if (page.locator && await page.locator(sel).first().count() > 0) {
                    return WafChallengeType.CLOUDFLARE_TURNSTILE;
                }
            }

            const title = typeof page.title === 'function' ? (await page.title()).toLowerCase() : '';
            if (title.includes('just a moment') || title.includes('attention required')) {
                return WafChallengeType.CLOUDFLARE_MANAGED;
            }

            for (const sel of HCAPTCHA_SELECTORS) {
                if (page.locator && await page.locator(sel).first().count() > 0) {
                    return WafChallengeType.HCAPTCHA;
                }
            }

            for (const sel of RECAPTCHA_SELECTORS) {
                if (page.locator && await page.locator(sel).first().count() > 0) {
                    return WafChallengeType.RECAPTCHA;
                }
            }
        } catch (e) {
            // Non-fatal inspection error
        }

        return WafChallengeType.NONE;
    }

    /**
     * Automatically attempts to click the Turnstile checkbox via human mouse physics.
     * @param {any} page
     * @param {Object} [options]
     * @param {HumanMouse} [options.humanMouse]
     * @param {number} [options.timeoutMs=12000]
     * @returns {Promise<boolean>}
     */
    static async solveTurnstile(page, options = {}) {
        const mouse = options.humanMouse || new HumanMouse(page);
        const timeoutMs = options.timeoutMs || 12000;
        const deadline = Date.now() + timeoutMs;

        while (Date.now() < deadline) {
            try {
                for (const frameSel of CLOUDFLARE_SELECTORS) {
                    if (page.locator && await page.locator(frameSel).first().count() > 0) {
                        const box = await page.locator(frameSel).first().boundingBox();
                        if (box && box.width > 20 && box.height > 20) {
                            const targetX = box.x + Math.min(36.0, box.width * 0.2) + (Math.random() * 8 - 4);
                            const targetY = box.y + (box.height * 0.5) + (Math.random() * 8 - 4);

                            await mouse.click({ x: targetX, y: targetY }, {
                                steps: Math.floor(Math.random() * 14) + 25,
                                jitter: true,
                                overshoot: true
                            });

                            await sleep(2000);
                            return true;
                        }
                    }
                }
            } catch (e) {
                // Retry next tick
            }

            await sleep(500);
        }

        return false;
    }
}

module.exports = {
    WafChallengeType,
    TurnstileEvaluator,
    CLOUDFLARE_SELECTORS,
    HCAPTCHA_SELECTORS,
    RECAPTCHA_SELECTORS
};
